versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/TestMongoDatabaseAdapter.java

```diff
@@ -17,100 +17,17 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-import de.flapdoodle.embed.mongo.Command;
-import de.flapdoodle.embed.mongo.MongodExecutable;
-import de.flapdoodle.embed.mongo.MongodStarter;
-import de.flapdoodle.embed.mongo.config.Defaults;
-import de.flapdoodle.embed.mongo.config.MongodConfig;
-import de.flapdoodle.embed.mongo.config.MongodProcessOutputConfig;
-import de.flapdoodle.embed.mongo.config.Net;
-import de.flapdoodle.embed.mongo.distribution.Version;
-import de.flapdoodle.embed.process.config.io.ProcessOutput;
-import de.flapdoodle.embed.process.io.StreamProcessor;
-import de.flapdoodle.embed.process.runtime.Network;
-import java.io.IOException;
-import java.util.concurrent.atomic.AtomicInteger;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
-import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
 import org.projectnessie.versioned.persist.tests.AbstractTieredCommitsTest;
 
+@ExtendWith(MongoDatabaseAdapterExtension.class)
 public class TestMongoDatabaseAdapter extends AbstractTieredCommitsTest {
```

Comment:
Just as an idea for consideration: We could use two extensions: one for managing the storage (MongoDB, RockDB, containers, etc) and another for managing adapters. That would allow tests for pick and choose adapter-storage permutations.

Also, since a lot of the tests are written in an adapter-neutral way, we could reuse the test classes for all adapter without requiring test sub-classes. That would allow us to bring nested tests to the top level to make them executable directly from the IDE.

Unfortunately, it does not seem possible to produce test contexts in JUnit5 from an extension without interfering with `@ParameterizedTest`. That is an extension can produce new test context for different adapter names and storage settings, but that will only work with simple `@Test` methods. `@ParameterizedTest` will not be able to resolve their parameters.

However, if we introduced a custom JUnit5 test _engine_ we would be able to create new branches in the test context hierarchy that would be compatible with `@ParameterizedTest`. Our own extensions would use the `UniqueId` segment information for figuring out storage settings and adapter names.

For example: currently a test with `UniqueId` of `[engine:junit-jupiter]/[class:org.projectnessie.versioned.persist.rocks.TestTieredCommitsRocks]/[nested-class:Concurrency]/[test-template:concurrency(org.projectnessie.versioned.persist.tests.AbstractConcurrency$Variation)]/[test-template-invocation:#1]`

Would become something like `[engine:junit-jupiter]/[storage:rocks-local][adapter:RocksDB][class:org.projectnessie.versioned.persist.rocks.TestTieredCommitsRocks]/[nested-class:Concurrency]/[test-template:concurrency(org.projectnessie.versioned.persist.tests.AbstractConcurrency$Variation)]/[test-template-invocation:#1]`

I have a sketch of this approach at: https://github.com/dimas-b/stargate/commit/40a88a4f9e9950338c38794915b2d68cbf3ee639 (cf. `StorageAwareEngine`)

Some highlights:
* one set of tests for all adapters would live in a separate maven module
* each adapter module will depend on common tests and define resources (or annotations) for discovering their specific database config settings
* The custom test engine will use annotations to find related tests and customize the text context tree for them during the JUnit 5 discovery phase
* IDEs will be able to run individual tests from source and re-run / debug specific test configurations from the test run pane by virtue of passing a specific test `UniqueId` to JUnit 5
* Maven will run all tests with all database configrations by introspecting test dependencies
* Our custom test _extensions_ will be able to injects relevant database / adapter objects based on `UniqueId` segments and test annotations.

Reply:
Not sure whether I'd like to go down the route of implementing a JUnit5 test engine, even if delegating to the "standard" one. I'd rather pay the price of adding "mostly empty" test classes at the moment.

It's not too hard to inject adapters with custom configurations into specific test cases from an extension, if the `DatabaseAdapter` (or `VersionStore`) is injected as a test method parameter. Not sure whether `@ParameterizedTest` really doesn't work with `org.junit.jupiter.api.extension.ParameterResolver` if the parameterized-test is annotated with the "right" `@ExtendWith`.